---
title: "Single Cell Data Simulation"
author: "Sushant Patkar"
date: "11/12/2019"
output: 
  md_document:
    variant: markdown_github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
#load required libraries
library(splatter)
library(scater)

#Enter destination path here
PATH = "~/Single Cell/forKun"

#helper function to assign cells to patients in a probablistic fashion 
assign_patients <- function(tpm, batch, celltype, fracs, n = 50){
  patient = rep(NA, length(batch))
  id = 0
  for(b in seq(length(unique(batch)))){
  for(i in seq(id + 1, id + round(length(batch)/(length(unique(batch))*length(unique(celltype))*n)))){
    print(i)
    ncells <- as.numeric(t(rmultinom(1, size = n, prob = fracs)))
    for(j in seq(length(fracs))){
      idx = which(is.na(patient))
      k = sample(which(batch[idx] == paste("Condition ",b,sep = "") & celltype[idx] == paste("celltype ",j, sep = "")), ncells[j])
      patient[idx][k] = paste("patient",i,sep = "")
    }
    id = i
  }
    
  }
  
  return(patient)
  
}

#helper function to generate ground truth datasets for deconvolution
generate_gt <- function(dataset,id, path, leave_out = 10){
cctypes = unique(dataset$celltype)
#obtain patient specific cell fractions
sim_cell_fracs <- sapply(unique(dataset$patients), function(x) {
  idx = which(dataset$patients == x)
  counts = sapply(cctypes, function(y) {
    return(sum(dataset$patients == x & dataset$celltype == y))
  })
  counts = counts/sum(counts)
  names(counts) = cctypes
  return(counts)
})

#obtain patient specific mean expression in each cell type
sim_ct_expr <- lapply(unique(dataset$patients), function(x) {
  cdat <- sapply(cctypes, function(y) {
    if(sum(dataset$patients == x & dataset$celltype == y) > 1){
    return(rowMeans(dataset$tpm[,dataset$patients == x & dataset$celltype == y],na.rm = T))
    }
    else if(sum(dataset$patients == x & dataset$celltype == y) == 1){
      return(dataset$tpm[,dataset$patients == x & dataset$celltype == y])
    }
    else{
      return(rep(0, nrow(dataset$tpm)))
    }
  })
  return(cdat)
})
names(sim_ct_expr) <- unique(dataset$patients)

sim_ct_expr2 <- lapply(cctypes, function(x){
  return(sapply(unique(dataset$patients), function(y){
    return(sim_ct_expr[[y]][,x])
  }))
})
names(sim_ct_expr2) = cctypes

#generate expected bulk expression profiles for each patient based on the single cell expression profile of each cell type in that patient
sim_bulk <- sapply(unique(dataset$patients), function(x) {
  bb <-sim_ct_expr[[x]]%*%sim_cell_fracs[,x]
})
rownames(sim_bulk) = rownames(dataset$tpm)
sim_cell_fracs <- t(sim_cell_fracs)

#set aside single cell data of 10 patients for signature:
set.seed(123)
left_out <- sample(colnames(sim_bulk), 10)
print(left_out)
rest <- sapply(colnames(sim_bulk), function(pt) !(pt %in% left_out))
idx <- sapply(dataset$patients, function(pt) pt %in% left_out)
sim_meta <- data.frame(Cell = colnames(dataset$tpm), celltype = dataset$celltype, patient = dataset$patients, batch = dataset$batch)

sc_sig <- dataset$tpm[,idx]
colnames(sc_sig) <- dataset$celltype[idx]
write.table(sc_sig, file = paste(path,"/Sim_singleCell_TPM_for_sig",id,".txt",sep = ""), sep = "\t", quote = F,col.names = NA)
saveRDS(dataset$tpm,file = paste(path,"/Sim_singleCell_TPM_",id,".rds",sep = ""))

#cell fraction estimates for cell types
saveRDS(sim_cell_fracs[rest,], file = paste(path, "/sim_2cell_cellfracs_",id,".rds",sep = ""))

sim_ct_expr2 <- lapply(sim_ct_expr2, function(x) {
  return(x[,rest])
})
#cell type specific expression in each patient: ground truth for deconvolution validation
saveRDS(sim_ct_expr2, file = paste(path, "/sim_2cell_groundtruth_tpm_",id,".rds",sep = ""))

#Livnat dataset updated cell --> cell type --> patient mapping
saveRDS(sim_meta[idx,], file = paste(path, "/sim_2cell_celltypes_table_",id,".rds",sep = ""))


#in-silico bulk gene expression for each of the 100 patients
saveRDS(sim_bulk[,rest], file = paste(path, "/sim_2cell_bulk_tpm_",id,".rds",sep = ""))
}

#cell types (M): (Celltype 1, Celltype 2, Celltype 3, Celltype 4)
#conditions (N): (A,B)
#Celltype 1 is least differentially expressed between conditions
#Celltype 4 is most differentially expressed between conditions

#Simulation dataset 1: Cell type 1 is most abundant cell type in the population of patients and celltype 4 is least abundant
#Simulation dataset 2: Cell type 4 is most abundant cell type in the population of patients and celltype 1 is least abundant

N <- 4
M <- 2
nCells <- 2500
nGenes <- 500
#effect size parameters controls distance of clusters
effect_size <- c(0.001,0.03,0.06,0.1)
params = list()
c <- list()
tpms <- list()
ctype <- c()
condition <- c()

for(i in seq(N)){
  cond <- sapply(seq(M), function(x) paste("Condition ",x,sep = ""))
  params[[i]] <- newSplatParams()
  params[[i]] <- setParams(params[[i]], update = list(mean.rate = 0.5, mean.shape = 0.5, nGenes = nGenes))
  c[[i]] <- splatSimulate(params[[i]], batchCells = rep(nCells,M),
                      batch.facLoc = effect_size[i], batch.facScale = effect_size[i],
                      verbose = FALSE)
  
  tpm(c[[i]]) <- calculateTPM(c[[i]], rowData(c[[1]])$Length)
  tpms[[i]] <- tpm(c[[i]])
  ctype <- c(ctype, rep(paste("celltype ",i,sep = ""), ncol(tpms[[i]])))
  condition <- c(condition, cond[as.numeric(sapply(as.character(c[[i]]$Batch), function(x) gsub("Batch","",x)))])
}
dat <- Reduce(cbind, tpms)


#Simulation dataset1: 
patients <- assign_patients(dat, condition, ctype, c(0.75,0.1,0.1,0.05))
dataset1 <- list()
dataset1$tpm <- dat[,!is.na(patients)]
dataset1$patients <- patients[!is.na(patients)]
dataset1$celltype <- as.character(ctype)
dataset1$celltype = dataset1$celltype[!is.na(patients)]
dataset1$batch = as.character(condition)
dataset1$batch = dataset1$batch[!is.na(patients)]
dataset1$sc_reference <- dat[,!is.na(patients)]



idx <- sample(ncol(dataset1$tpm),1000)
res <- prcomp(t(dataset1$tpm[rowSums(dataset1$tpm[,idx]) != 0, idx]),scale = T)
print(dim(res$x))
p1 <- ggplot(data = data.frame(dim1 = res$x[,1],dim2 = res$x[,2], condition = dataset1$batch[idx], celltype = dataset1$celltype[idx]), aes(x= dim1, y = dim2, color = condition, shape = celltype)) + geom_point() + theme_bw()
print(p1)
ggsave(p1, filename = paste(PATH,"/sim_dataset1_pca.pdf",sep = ""), useDingbats = F)
generate_gt(dataset1,1, PATH)
rm(dataset1)


#Simulation dataset2: 
patients <- assign_patients(dat, condition, ctype, c(0.05,0.1,0.1,0.75))
dataset2 <- list()
dataset2$tpm <- dat[,!is.na(patients)]
dataset2$patients <- patients[!is.na(patients)]
dataset2$celltype <- as.character(ctype)
dataset2$celltype = dataset2$celltype[!is.na(patients)]
dataset2$batch = as.character(condition)
dataset2$batch = dataset2$batch[!is.na(patients)]
rm(dat)


idx <- sample(ncol(dataset2$tpm),1000)
res <- prcomp(t(dataset2$tpm[rowSums(dataset2$tpm[,idx]) != 0, idx]),scale = T)
print(dim(res$x))
p2 <- ggplot(data = data.frame(dim1 = res$x[,1],dim2 = res$x[,2], condition = dataset2$batch[idx], celltype = dataset2$celltype[idx]), aes(x= dim1, y = dim2, color = condition, shape = celltype)) + geom_point() + theme_bw()
print(p2)
ggsave(p2, filename = paste(PATH,"/sim_dataset2_pca.pdf",sep = ""), useDingbats = F)
generate_gt(dataset2,2, PATH)
rm(dataset2)

rm(tpms)
rm(res)
rm(c)

####################################################################################################################################
#cell types (M): (Celltype 1, Celltype 2, Celltype 3, Celltype 4)
#conditions (N): (1,2,3)
#Celltype 1 is least differentially expressed between conditions
#Celltype 4 is most differentially expressed between conditions

#Simulation dataset 3: Cell type 1 is most abundant cell type in the population of patients and celltype 4 is least abundant
#Simulation dataset 4: Cell type 4 is most abundant cell type in the population of patients and celltype 1 is least abundant
N <- 4
M <- 3
nCells <- 1700
nGenes <- 500
effect_size <- c(0.001,0.03,0.05,0.1)
params = list()
c <- list()
tpms <- list()
ctype <- c()
condition <- c()
for(i in seq(N)){
  cond <- sapply(seq(M), function(x) paste("Condition ",x,sep = ""))
  params[[i]] <- newSplatParams()
  params[[i]] <- setParams(params[[i]], update = list(mean.rate = 0.5, mean.shape = 0.5, nGenes = nGenes))
  c[[i]] <- splatSimulate(params[[i]], batchCells = rep(nCells,M),
                      batch.facLoc = effect_size[i], batch.facScale = effect_size[i],
                      verbose = FALSE)
  
  tpm(c[[i]]) <- calculateTPM(c[[i]], rowData(c[[1]])$Length)
  tpms[[i]] <- tpm(c[[i]])
  ctype <- c(ctype, rep(paste("celltype ",i,sep = ""), ncol(tpms[[i]])))
  condition <- c(condition, cond[as.numeric(sapply(as.character(c[[i]]$Batch), function(x) gsub("Batch","",x)))])
}
dat <- Reduce(cbind, tpms)


#Simulation dataset3: 
patients <- assign_patients(dat, condition, ctype, c(0.75,0.1,0.1,0.05))
dataset3 <- list()
dataset3$tpm <- dat[,!is.na(patients)]
dataset3$patients <- patients[!is.na(patients)]
dataset3$celltype <- as.character(ctype)
dataset3$celltype = dataset3$celltype[!is.na(patients)]
dataset3$batch = as.character(condition)
dataset3$batch = dataset3$batch[!is.na(patients)]



idx <- sample(ncol(dataset3$tpm),1000)
res <- prcomp(t(dataset3$tpm[rowSums(dataset3$tpm[,idx]) != 0, idx]),scale = T)
print(dim(res$x))
p3 <- ggplot(data = data.frame(dim1 = res$x[,1],dim2 = res$x[,2], condition = dataset3$batch[idx], celltype = dataset3$celltype[idx]), aes(x= dim1, y = dim2, color = condition, shape = celltype)) + geom_point() + theme_bw()
print(p3)
ggsave(p3, filename = paste(PATH, "/sim_dataset3_pca.pdf", sep = ""), useDingbats = F)
generate_gt(dataset3,3, PATH)
rm(dataset3)


#Simulation dataset4: 
patients <- assign_patients(dat, condition, ctype, c(0.05,0.1,0.1,0.75))
dataset4 <- list()
dataset4$tpm <- dat[,!is.na(patients)]
dataset4$patients <- patients[!is.na(patients)]
dataset4$celltype <- as.character(ctype)
dataset4$celltype = dataset4$celltype[!is.na(patients)]
dataset4$batch = as.character(condition)
dataset4$batch = dataset4$batch[!is.na(patients)]
rm(dat)


idx <- sample(ncol(dataset4$tpm),1000)
res <- prcomp(t(dataset4$tpm[rowSums(dataset4$tpm[,idx]) != 0, idx]),scale = T)
print(dim(res$x))
p4 <- ggplot(data = data.frame(dim1 = res$x[,1],dim2 = res$x[,2], condition = dataset4$batch[idx], celltype = dataset4$celltype[idx]), aes(x= dim1, y = dim2, color = condition, shape = celltype)) + geom_point() + theme_bw()
print(p4)
ggsave(p4, filename = paste(PATH, "/sim_dataset4_pca.pdf", sep = ""), useDingbats = F)
generate_gt(dataset4,4, PATH)
rm(dataset4)
```

